*) Fix a typo in hints/aix.sh.
*) Need to pass linker flags through gcc.
*) Build the shared object using gcc flag '-shared' and linker flag '-G'.
*) Give the shared object packed into libperl.a some useful soname.

--- hints/aix.sh
+++ hints/aix.sh
@@ -295,10 +295,10 @@
 	    fi
 	if test X"$use64bitint:$quadtype" = X"$define:long" -o X"$use64bitall" = Xdefine; then
 # Keep this at the left margin.
-libswanted_uselargefiles="`getconf XBS5_LP64_OFF64_LIBS 2>/dev/null|sed -e 's@^-l@@' -e 's@ -l@ @g`"
+libswanted_uselargefiles="`getconf XBS5_LP64_OFF64_LIBS 2>/dev/null|sed -e 's@^-l@@' -e 's@ -l@ @g'`"
 	else
 # Keep this at the left margin.
-libswanted_uselargefiles="`getconf XBS5_ILP32_OFFBIG_LIBS 2>/dev/null|sed -e 's@^-l@@' -e 's@ -l@ @g`"
+libswanted_uselargefiles="`getconf XBS5_ILP32_OFFBIG_LIBS 2>/dev/null|sed -e 's@^-l@@' -e 's@ -l@ @g'`"
 	    fi
 
 	case "$ccflags_uselargefiles$ldflags_uselargefiles$libs_uselargefiles" in
@@ -346,7 +346,7 @@
 		ccflags="`echo $ccflags | sed -e 's@ -b@ -Wl,-b@g'`"
 		ldflags="`echo ' '$ldflags | sed -e 's@ -b@ -Wl,-b@g'`"
 		lddlflags="`echo ' '$lddlflags | sed -e 's@ -b@ -Wl,-b@g'`"
-		lddlflags="`echo ' '$lddlflags | sed -e 's@ -G @ -Wl,-G @g'`"
+		lddlflags="`echo ' '$lddlflags | sed -e 's@ -G @ -shared -Wl,-G @g'`"
 
 	    	# Due to calling $cc without $cflags when linking some
 		# binaries we need to hardwire $cc to the right mode.
--- Makefile.SH
+++ Makefile.SH
@@ -81,11 +81,15 @@
 		shrpldflags="$shrpldflags -Wl,--image-base,0x57000000"
 		;;
 	aix*)
-		shrpldflags="-H512 -T512 -bhalt:4 -bM:SRE -bE:perl.exp"
+		case "$cc" in
+		*gcc*) wl='-Wl,'; shrpldflags="-shared ${wl}-G" ;;
+		*)     wl='';     shrpldflags="-G" ;;
+		esac
+		shrpldflags="${shrpldflags} ${wl}-H512 ${wl}-T512 ${wl}-bhalt:4 ${wl}-bernotok ${wl}-bE:perl.exp"
 		case "$osvers" in
-		3*)	shrpldflags="$shrpldflags -e _nostart"
+		3*)	shrpldflags="$shrpldflags ${wl}-e _nostart"
 			;;
-		*)	shrpldflags="$shrpldflags -bnoentry"
+		*)	shrpldflags="$shrpldflags ${wl}-bnoentry"
 			;;
 		esac
 	        shrpldflags="$shrpldflags $ldflags $perllibs $cryptlib"
@@ -598,9 +602,9 @@
 		case "$osname" in
 		aix)
 			$spitshell >>Makefile <<'!NO!SUBS!'
-	rm -f libperl$(OBJ_EXT)
-	mv $@ libperl$(OBJ_EXT)
-	$(AR) qv $(LIBPERL) libperl$(OBJ_EXT)
+	v=$(LIBPERL_SOVERSION); rm -f libperl$(OBJ_EXT)$${v:+.$${v}}
+	v=$(LIBPERL_SOVERSION); mv $@ libperl$(OBJ_EXT)$${v:+.$${v}}
+	v=$(LIBPERL_SOVERSION); $(AR) qv $(LIBPERL) libperl$(OBJ_EXT)$${v:+.$${v}}
 !NO!SUBS!
 			;;
 		esac
